### PR TITLE
2598: Fix follow user location and decouple distance from map features (native)

### PR DESCRIPTION
**Pull request by steffenkleinle**
### Short description

This fixes the bug that the map always centers back on a selected map feature instead of keeping the manual set viewport or staying on the users location.
This was caused by continuous rerenders triggered by very frequent updates of the user location which made the whole Pois route rerender (and therefore trigger the useEffect responsible for centering on a newly selected map feature). To really fix this, sadly some refactoring was necessary which I split in web and native.

### Proposed changes

- Decouple pois and map features (and get rid of usages of GeoJsonPois)
- Remove distance from map features (as well as other properties not needed for displaying map features) and calculate when needed
- Only update distance when diff greater than threshold (0.1km)
- Extract and reuse filtering/multipoi/slug logic
- Some smaller refactoring to separate concerns a little better

### Side effects

Web will only work once the corresponding PR is merged (#2641).

### Testing

Please test this thoroughly. Especially please test selecting/deselecting multipois (and what is displayed in the bottom action sheet). Also check whether scroll positions are kept.

### Resolved issues

Fixes: #2598 (native).

---

### CODE REVIEW

File: native/src/components/PoiDetails.tsx [hunk 1]
```diff
@@ -44,14 +43,14 @@ const StyledCategory = styled.Text`
...
+  distance: number | null
 }
 
-const PoiDetails = ({ poi, poiFeature, language }: PoiDetailsProps): ReactElement => {
+const PoiDetails = ({ poi, language, distance }: PoiDetailsProps): ReactElement => {
   const { t } = useTranslation('pois')
-  const thumbnail = poiFeature.thumbnail?.replace('-150x150', '') ?? PoiThumbnailPlaceholderLarge
-  const distance = poiFeature.distance
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  const thumbnail = poi.thumbnail?.replace('-150x150', '') ?? PoiThumbnailPlaceholderLarge
```
Comment (steffenkleinle) [inline native/src/components/PoiDetails.tsx:53]: Sadly our types here are wrong, `thumbnail` may very well be null (see https://digitalfabrik.github.io/integreat-cms/api-docs.html#locations).

Comment (LeandraH) [reply native/src/components/PoiDetails.tsx:53]: Would it make sense to open a ticket to fix them?

Comment (steffenkleinle) [reply native/src/components/PoiDetails.tsx:53]: Ahh sorry, included it in this issue: #2635.

File: api-client/src/utils/geoJson.ts [hunk 2]
```diff
@@ -1,7 +1,9 @@
 import distance from '@turf/distance'
 
-import { featureLayerId, LocationType, PoiModel } from '../index'
-import { GeoJsonPoi, MapFeature, MapFeatureCollection } from '../maps'
+import { featureLayerId, LocationType, MapFeature, MapFeatureCollection } from '../maps'
+import PoiModel from '../models/PoiModel'
+
+export const MIN_DISTANCE_THRESHOLD = 0.1
```
[the remark this answers is not in the document]
Comment (steffenkleinle) [reply api-client/src/utils/geoJson.ts:6]: Thats the precision that is atm displayed to the user. We could of course change this if we decide it should be more precise.

**Review by f1sh1918 — COMMENTED** [review]
i'm facing a super strange issue on ios when filtering pois, just this particular category.
Can anyone test on ios if this is just a simulator issue
On android everything works fine

https://github.com/digitalfabrik/integreat-app/assets/37902063/d9af2e6e-0fa2-466d-b0c9-eccd2f995498

**Review by f1sh1918 — CHANGES_REQUESTED** [review]
I'm facing a super strange rendering issue of `PoiListItem`
1. When filtering pois, just this particular category. Only happens if you didn't filter or select anything before.
2. Selecting a multipoi results in the same issue

Happens on ios (simulators and real device)
On android everything works fine

I played around with the style changes but couldn't find the issue so far.
On main everything is fine, so it is definitly related to this pr

https://github.com/digitalfabrik/integreat-app/assets/37902063/d9af2e6e-0fa2-466d-b0c9-eccd2f995498

https://github.com/digitalfabrik/integreat-app/assets/37902063/f40fae34-9bbe-4e5f-8f18-7271dcafc75c

File: native/src/components/PoiListItem.tsx [hunk 3]
```diff
@@ -28,15 +28,14 @@ const Thumbnail = styled(SimpleImage)`
 
 const StyledPressable = styled(Pressable)<{ language: string }>`
   flex: 1;
+  height: 100%;
```
Comment (f1sh1918) [inline native/src/components/PoiListItem.tsx:31]: this causes rendering issues on ios. please remove `flex: 1; height: 100%`

File: native/src/components/PoiListItem.tsx [hunk 4]
```diff
@@ -28,15 +28,14 @@ const Thumbnail = styled(SimpleImage)`
...
   border-bottom-width: 1px;
   border-bottom-color: ${props => props.theme.colors.textDisabledColor};
   flex-direction: ${props => contentDirection(props.language)};
   padding: 24px 0;
 `
 
 const Description = styled.View`
-  flex: 1;
```
Comment (f1sh1918) [inline native/src/components/PoiListItem.tsx:38]: readd `flex:1`

**Comment by f1sh1918** [comment]
@steffenkleinle please adjust the styles for `PoiListItem` as following to solve the ios rendering issues

```js
const StyledPressable = styled(Pressable)<{ language: string }>`
  border-bottom-width: 1px;
  border-bottom-color: ${props => props.theme.colors.textDisabledColor};
  flex-direction: ${props => contentDirection(props.language)};
  padding: 24px 0;
`

const Description = styled.View`
  flex: 1;
  flex-direction: column;
  font-family: ${props => props.theme.fonts.native.decorativeFontRegular};
  padding: 0 32px;
  justify-content: center;
`
```
Can not find out why you added a height and so one. We just need a flex on Description

**Review by f1sh1918 — APPROVED** [review]

File: native/src/components/List.tsx [hunk 5]
```diff
@@ -44,11 +44,7 @@ const List = <T,>({
     }
     onEndReached={onEndReached}
     showsVerticalScrollIndicator={false}
-    contentContainerStyle={{
-      flexGrow: 1,
-
-      paddingHorizontal: 10,
-    }}
+    contentContainerStyle={{ flex: 1 }}
```
Comment (ztefanie) [inline native/src/components/List.tsx:47]: @steffenkleinle Why have you changed this? This causes this bug: https://github.com/digitalfabrik/integreat-app/issues/2660

I reverted here: Feel free to comment if this breaks something you fixed.

Comment (steffenkleinle) [reply native/src/components/List.tsx:47]: Hmm, I think I had some layout problems, but @f1sh1918 suggestions already fixed them. So yes, feel free to revert.

**Pull request merged**